src/passes/MultiMemoryLowering.cpp

```diff
@@ -0,0 +1,345 @@
+        offsetGlobalNames.push_back(std::move(name));
+        addGlobal(name, offsetRunningTotal);
+      }
+      offsetRunningTotal += memory->initial;
+    }
+  }
+
+  void adjustActiveDataSegmentOffsets() {
+    Builder builder(*wasm);
+    ModuleUtils::iterActiveDataSegments(*wasm, [&](DataSegment* dataSegment) {
+      assert(dataSegment->offset->is<Const>());
+      auto iter = memoryIdxMap.find(dataSegment->memory);
+      assert(iter != memoryIdxMap.end());
+      dataSegment->memory = combinedMemory;
+      // No need to update the offset of data segments for the first memory
+      if (iter->second != 0) {
+        // We need to subtract 1 here because there is no offsetGlobal for the
+        // first memory
+        auto offsetGlobalName = offsetGlobalNames[iter->second - 1];
+        dataSegment->offset = builder.makeBinary(
+          pointerType == Type::i32 ? AddInt32 : AddInt64,
+          builder.makeGlobalGet(offsetGlobalName, pointerType),
+          dataSegment->offset);
+      }
+    });
+  }
+
+  void createMemorySizeFunctions() {
+    // Don't create a memory size function for the last memory in the vector as
+    // its the combinedMemory we just added
+    for (Index i = 0; i < wasm->memories.size() - 1; i++) {
+      auto function = memorySize(i);
+      memorySizeNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  void createAdjustOffsetGlobalFunctions() {
+    // Don't create a memory size function for the last memory in the vector as
+    // its the combinedMemory we just added
+    for (Index i = 0; i < wasm->memories.size() - 1; i++) {
+      auto function = adjustOffsetGlobals(i);
+      adjustOffsetNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  // This function replaces memory.grow instruction calls in the wasm module.
+  // Because the multiple discrete memories are lowered into a single memory,
+  // we need to adjust offsets as a particular memory receives an
+  // instruction to grow.
+  std::unique_ptr<Function> adjustOffsetGlobals(Index memIdx) {
+    Builder builder(*wasm);
+    Name functionName =
+      Names::getValidFunctionName(*wasm, "adjust_memory_offsets");
+    auto function = Builder::makeFunction(
+      functionName, Signature(pointerType, pointerType), {});
+    function->hasExplicitName = true;
+    function->setLocalName(0, "page_delta");
+    auto getPageDelta = [&]() { return builder.makeLocalGet(0, pointerType); };
+    Expression* functionBody;
+
+    Index returnLocal = Builder::addVar(function.get(), pointerType);
+    functionBody = builder.blockify(builder.makeLocalSet(
+      returnLocal, builder.makeCall(memorySizeNames[memIdx], {}, pointerType)));
+
+    // If we are not growing the last memory, then we need to copy data,
+    // shifting it over to accomodate the increase from page_delta
+    if (memIdx != offsetGlobalNames.size()) {
```

Comment:
Since this also depends on the detail that `offsetGlobalNames` contains N-1 memories, it would be good to make a helper function like `isLastMemory` for this.